# Machine Learning Engineer Nanodegree
## Supervised Learning
## Project 2: Building a Student Intervention System

Welcome to the second project of the Machine Learning Engineer Nanodegree! In this notebook, some template code has already been provided for you, and it will be your job to implement the additional functionality necessary to successfully complete this project. Sections that begin with **'Implementation'** in the header indicate that the following block of code will require additional functionality which you must provide. Instructions will be provided for each section and the specifics of the implementation are marked in the code block with a `'TODO'` statement. Please be sure to read the instructions carefully!

In addition to implementing code, there will be questions that you must answer which relate to the project and your implementation. Each section where you will answer a question is preceded by a **'Question X'** header. Carefully read each question and provide thorough answers in the following text boxes that begin with **'Answer:'**. Your project submission will be evaluated based on your answers to each of the questions and the implementation you provide.  

>**Note:** Code and Markdown cells can be executed using the **Shift + Enter** keyboard shortcut. In addition, Markdown cells can be edited by typically double-clicking the cell to enter edit mode.

### Question 1 - Classification vs. Regression
*Your goal for this project is to identify students who might need early intervention before they fail to graduate. Which type of supervised learning problem is this, classification or regression? Why?*

Classification, as we have two discrete classes for various outcomes as comapared to continous outputs for regression.  

## Exploring the Data
Run the code cell below to load necessary Python libraries and load the student data. Note that the last column from this dataset, `'passed'`, will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.

In [2]:
# Import libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from time import time
from sklearn.metrics import f1_score
from pandas.tools.plotting import scatter_matrix
from sklearn import preprocessing
from IPython.display import display
from statsmodels.graphics.mosaicplot import mosaic

# Read student data
student_data = pd.read_csv("student-data.csv")
#print(student_data.head(2))


def handle_non_numeric_data(df):
    columns=df.columns.values
    for column in columns:
       if (df[column].dtype != np.int64 or df[column].dtype != np.float):
        le = preprocessing.LabelEncoder()
        le.fit(df[column])
        list(le.classes_)
        df[column]=le.transform(df[column])
        #print "Column", column, "List", list(le.classes_)

    return

#display(student_data)

print "Student data read successfully!"

# read data in plot_student_data for plotting in 


feature_cols = list(student_data.columns[:-1])
print feature_cols
plt.rcParams['font.size'] = 8.0


#for i, feature in enumerate(feature_cols):
#    print i ,feature
#    mosaic(student_data, [feature,'passed']);
#    #student_data.groupby(feature).size().plot(kind='bar')
#    plt.xlabel(feature)
#    plt.show()

Student data read successfully!
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


### Implementation: Data Exploration
Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
- The total number of students, `n_students`.
- The total number of features for each student, `n_features`.
- The number of those students who passed, `n_passed`.
- The number of those students who failed, `n_failed`.
- The graduation rate of the class, `grad_rate`, in percent (%).


In [3]:
#print student_data.head(2)
#features=student_data.drop(['passed'],axis=1)
#result=student_data['passed']

#print features.head(2)
#print result.shape[0] 

# TODO: Calculate number of students
n_students = student_data.shape[0] 

# TODO: Calculate number of features
#n_features = features.shape[1]
n_features = student_data.shape[1] -1


# TODO: Calculate passing students
n_passed = n_students *student_data['passed'].value_counts('yes')[0]

# TODO: Calculate failing students
n_failed = n_students*student_data['passed'].value_counts('no')[1]

# TODO: Calculate graduation rate
grad_rate = student_data['passed'].value_counts('yes')[0]*100.0

# Print the results
print "Total number of students: {}".format(n_students)
print "Number of features: {}".format(n_features)
print "Number of students who passed: {}".format(n_passed)
print "Number of students who failed: {}".format(n_failed)
print "Graduation rate of the class: {:.2f}%".format(grad_rate)


    

Total number of students: 395
Number of features: 30
Number of students who passed: 265.0
Number of students who failed: 130.0
Graduation rate of the class: 67.09%


## Preparing the Data
In this section, we will prepare the data for modeling, training and testing.

### Identify feature and target columns
It is often the case that the data you obtain contains non-numeric features. This can be a problem, as most machine learning algorithms expect numeric data to perform computations with.

Run the code cell below to separate the student data into feature and target columns to see if any features are non-numeric.

In [4]:
# Extract feature columns
feature_cols = list(student_data.columns[:-1])

# Extract target column 'passed'
target_col = student_data.columns[-1] 

# Show the list of columns
print "Feature columns:\n{}".format(feature_cols)
print "\nTarget column: {}".format(target_col)

# Separate the data into feature data and target data (X_all and y_all, respectively)
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# Show the feature information by printing the first five rows
print "\nFeature values:"
print X_all.head(5)

print "\nTarget values:"
print y_all.head(5)

#le = preprocessing.LabelEncoder()
#le.fit(y_all)
#y_all=le.transform(y_all)


#X_all_copy=X_all

##Print features & classes of each feature
#for feature_type in feature_cols:
# le = preprocessing.LabelEncoder()
# le.fit(X_all_copy[feature_type])
# X_all_copy[feature_type]=le.transform(X_all_copy[feature_type])   
# print feature_type, list(le.classes_)
   
 
 


Feature columns:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Target column: passed

Feature values:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...       

### Preprocess Feature Columns

As you can see, there are several non-numeric columns that need to be converted! Many of them are simply `yes`/`no`, e.g. `internet`. These can be reasonably converted into `1`/`0` (binary) values.

Other columns, like `Mjob` and `Fjob`, have more than two values, and are known as _categorical variables_. The recommended way to handle such a column is to create as many columns as possible values (e.g. `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc.), and assign a `1` to one of them and `0` to all others.

These generated columns are sometimes called _dummy variables_, and we will use the [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.

In [5]:
# Preprocess feature columns
def preprocess_features(X):
    outX = pd.DataFrame(index=X.index)  # output dataframe, initially empty

    # Check each column
    for col, col_data in X.iteritems():
        # If data type is non-numeric, try to replace all yes/no values with 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])
        # Note: This should change the data type for yes/no columns to int

        # If still non-numeric, convert to one or more dummy variables
        if col_data.dtype == object:
            col_data = pd.get_dummies(col_data, prefix=col)  # e.g. 'school' => 'school_GP', 'school_MS'

        outX = outX.join(col_data)  # collect column(s) in output dataframe

    return outX

X_all = preprocess_features(X_all)
print "Processed feature columns ({}):-\n{}".format(len(X_all.columns), list(X_all.columns))
y_all = y_all.replace(['yes', 'no'], [1, 0]) # changed to enable f1 scoring during grid search

#feature_col=list(X_all.columns[:])
#print feature_col 
#for i, feature in enumerate(feature_col):
#    print i ,feature

#feature_cols = list(X_all.columns[:-1])

##for feature_type in feature_cols:
# le = preprocessing.LabelEncoder()
# le.fit(X_all[feature_type])
# print feature_type, list(le.classes_)
# X=X_all[feature_type];
# print X 
# index_pass=
# plt.scatter(X, y_all, marker=">")
# plt.show()   
   
#le = preprocessing.LabelEncoder()
#le.fit(y_all)
#list(le.classes_)
#y_all=le.transform(y_all)
#print y_all

#print "Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns))


Processed feature columns (48):-
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


### Implementation: Training and Testing Data Split
So far, we have converted all _categorical_ features into numeric values. For the next step, we split the data (both features and corresponding labels) into training and test sets. In the following code cell below, you will need to implement the following:
- Randomly shuffle and split the data (`X_all`, `y_all`) into training and testing subsets.
  - Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).
  - Set a `random_state` for the function(s) you use, if provided.
  - Store the results in `X_train`, `X_test`, `y_train`, and `y_test`.

In [15]:
# TODO: Import any additional functionality you may need here
from sklearn.cross_validation import train_test_split

# TODO: Set the number of training points
num_train = 0.8*X_all.shape[0]

# Set the number of testing points
num_test = X_all.shape[0] - num_train
test_size=num_test/(num_test+num_train)


X_train, X_test, y_train, y_test = train_test_split(X_all,y_all,test_size=test_size, random_state=10)
#X_train, X_test, y_train, y_test = train_test_split(X_all,y_all,test_size=test_size, stratify=y_all,random_state=10)


# Show the results of the split
print "Training set has {} samples.".format(X_train.shape[0])
print "Testing set has {} samples.".format(X_test.shape[0])

#print "\n X-train", X_train.head(2)
#print "\n X-test", X_test.head(2)
#print "\n Y-train", y_train.head(2)
#print "\n Y-test", y_test.head(2)

Training set has 316 samples.
Testing set has 79 samples.


## Training and Evaluating Models
In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in `scikit-learn`. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to varying sizes of training data (100 data points, 200 data points, and 300 data points) and measure the F<sub>1</sub> score. You will need to produce three tables (one for each model) that shows the training set size, training time, prediction time, F<sub>1</sub> score on the training set, and F<sub>1</sub> score on the testing set.

**The following supervised learning models are currently available in** [`scikit-learn`](http://scikit-learn.org/stable/supervised_learning.html) **that you may choose from:**
- Gaussian Naive Bayes (GaussianNB)
- Decision Trees
- Ensemble Methods (Bagging, AdaBoost, Random Forest, Gradient Boosting)
- K-Nearest Neighbors (KNeighbors)
- Stochastic Gradient Descent (SGDC)
- Support Vector Machines (SVM)
- Logistic Regression

### Question 2 - Model Application
*List three supervised learning models that are appropriate for this problem. For each model chosen*
- Describe one real-world application in industry where the model can be applied. *(You may need to do a small bit of research for this — give references!)* 
- What are the strengths of the model; when does it perform well? 
- What are the weaknesses of the model; when does it perform poorly?
- What makes this model a good candidate for the problem, given what you know about the data?

### I consider following three supervised learning models for this classification problem.

1. Decision Trees.
2. SVM.
3. Naive Bayes classification

# Decision trees 
Decision trees are non-parametrized supervised learning methods, can be used for both classification & regression.
This method is really good for cases where ability to explain the critieria for classification matters more than accuracy. For example in identifying customer segments from given marketing survey, it is important to have the classification knowledge in human usable form, which can be validated as per the experience of a professional in that area & practically used. This is different from the abstract classification done by the neural networks.

#### Strengths of Decision trees
1. Generate intutive rules for classification.
2. Computationaly inexpensive for classification of new test samples.
3. Best works for instances represented by (attribute,value) pair when value take discrete values. Example (attribute=Color, Value=red,yellow,white)
4. Can handle instances with continous values. Discretize based on thresholds that gives best classification of target attribute. 
5. Can handle well error in data. It uses all training examples at each step to refine current hypothesis, so less sensitive to error/noise in data.

#### Weaknesses of Decision trees
1. Can overfit the data, grow deep enough to classify each training example. 
 ..1.Remedy to stop growing beyond a depth. Generally not used as difficult to know when to stop growing the tree.
 ..2.Prune the tree. Best strategy, might lead to increased error. Pruning is guided by testing the pruned tree on     test data. Use pruning which leads to minimum loss in accuracy.
2. Can overfit if number of training samples small as compared to attributes. Not suitable for these problems.
3. Not good for predicitng the continous attributes.
4. Comptationally expensive to train O(N*V^2) where N=number of training examples, V=Number of features in a   
   training example. The training samples need to be sorted as each node is split. Pruning   
   algorithms are expensive, form & test new sub trees.
5. Classification regions are rectangular boxes, don't fit well to problems with non-rectangular classification  
   boxes.
     
The problem at hand is typical classification problem with discrete (attibute,value) pairs & discrete target attribute. Ratio of training samples to attribute is quite low, means there could be overfit proble. But we chose this startegy as it provide us visible intution on the most important attributes for clasiification. We could try to tackle the overfitting further using crossvalidation & pruning. 
     
Reference
1. Sklearn documentation.
2. Machine Learning by Tom Mitchell
3. Udacity  course notes
4. http://web.cs.ucdavis.edu/~vemuri/classes/ecs271/Decision%20Trees-Overview.htm


# Naive Bayes classifier

It's a probabilistic classifier based on assumption that all the attributes of training set are independent of each other. It finds application in text classification problems, where given instances of text documents they need to classified in various categories. This application doesn't really satisfy the naive independent criteria for attributes of the input instances. However as described in this paper (http://www.aaai.org/Papers/FLAIRS/2004/Flairs04-097.pdf), the uniform distribution of interdependence between attributes of input instance cancels resulting in good performance for text classification problems. However they might not perform good if interdependence is lop-sided towards few input attributes.    

#### Stengths of GNB
1. Require small amount of training data.
2. EXtremely fast due to assumption of decoupled feature distribution. Each feature distribution can be independently estimated. This alleviate the curse of dimensioanlity.
3. Can be updated on fly for each training data. good for streaming data.

#### Weaknesses
1. GNB assumes no dependence between attributes of training set.  
2 .Side-effect of above limitation it can’t learn interactions between features. 

We chose GNB as one option for the atudent classificatin problem due to its computation speed & abiltiy to work on small training set.

Reference
1. Sklearn documentation.
2. Machine Learning by Tom Mitchell
3. Udacity  course notes
4. http://web.cs.ucdavis.edu/~vemuri/classes/ecs271/Decision%20Trees-Overview.htm
5. stackflow
6. wiki

# SVM
Supervised learning algorithm used for classification & regression. The data points are p-dimensional vectors mapped to two different categories. SVM finds a linear classifier which is p-1 dimensional hyperplane taht maximizes the margin between two classes.

It is used for image classification problems, where mxn pixel image is a mxn dimensional input vector which is classified in two/more classes. Multi class classification is extension of binary SVM with one class vs all classes.


#### Advantages
1. effective in high dimensional space, large number of features.
2. effective even when number of dimensions greater than number of samples.
3. use subset of training data in decision making(one close to decision boundary) known as support vectors.Memory efficient in terms of storage.
4. Versatile for non-linear hyperplanes using kernel tricks.


#### Dis-advatnages
1. SVM is binary classification. Multi-class classification possible as eaxtension of one class vs. all other 
   classes classification.
2. Biased if kernel is incorrect
3. Computationally expensive.

SVM works great in high dimensional space. In our case features are less than number of samples but our data set contains lot of features relative to size of data-set. This is one main reason to chose SVM.


Reference
1. Sklearn documentation.
2. Udacity  course notes
3. stackflow
4. wiki




### Setup
Run the code cell below to initialize three helper functions which you can use for training and testing the three supervised learning models you've chosen above. The functions are as follows:
- `train_classifier` - takes as input a classifier and training data and fits the classifier to the data.
- `predict_labels` - takes as input a fit classifier, features, and a target labeling and makes predictions using the F<sub>1</sub> score.
- `train_predict` - takes as input a classifier, and the training and testing data, and performs `train_clasifier` and `predict_labels`.
 - This function will report the F<sub>1</sub> score for both the training and testing data separately.

In [16]:
from time import time
from sklearn import tree
from sklearn.externals.six import StringIO  
import pydotplus

def train_classifier(clf, X_train, y_train):
    ''' Fits a classifier to the training data. '''
    
    # Start the clock, train the classifier, then stop the clock
    start = time()
    clf.fit(X_train, y_train)
    end = time()
    
    # Print the results
    print "Trained model in {:.4f} seconds".format(end - start)

    
def predict_labels(clf, features, target):
    ''' Makes predictions using a fit classifier based on F1 score. '''
    
    # Start the clock, make predictions, then stop the clock
    start = time()
    y_pred = clf.predict(features)
    end = time()
    #print "\n target" ,target.values
    #print "\n prediction", y_pred
    
    # Print and return results
    print "Made predictions in {:.4f} seconds.".format(end - start)
    return f1_score(target.values, y_pred, pos_label=1)


def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' Train and predict using a classifer based on F1 score. '''
    
    # Indicate the classifier and the training set size
    print "Training a {} using a training set size of {}. . .".format(clf.__class__.__name__, len(X_train))
    
    # Train the classifier
    
    train_classifier(clf, X_train, y_train)
    #print clf
    if "Decision" in (clf.__class__.__name__):
     tree.export_graphviz(clf,out_file='tree.dot')    
     dot_data = StringIO() 
     tree.export_graphviz(clf, out_file=dot_data) 
     graph = pydotplus.graph_from_dot_data(dot_data.getvalue()) 
     graph.write_pdf("ouput.pdf")
    
    # Print the results of prediction for both training and testing
    print "F1 score for training set: {:.4f}.".format(predict_labels(clf, X_train, y_train))
    print "F1 score for test set: {:.4f}.".format(predict_labels(clf, X_test, y_test))
    




### Implementation: Model Performance Metrics
With the predefined functions above, you will now import the three supervised learning models of your choice and run the `train_predict` function for each one. Remember that you will need to train and predict on each classifier for three different training set sizes: 100, 200, and 300. Hence, you should expect to have 9 different outputs below — 3 for each model using the varying training set sizes. In the following code cell, you will need to implement the following:
- Import the three supervised learning models you've discussed in the previous section.
- Initialize the three models and store them in `clf_A`, `clf_B`, and `clf_C`.
 - Use a `random_state` for each model you use, if provided.
 - **Note:** Use the default settings for each model — you will tune one specific model in a later section.
- Create the different training set sizes to be used to train each model.
 - *Do not reshuffle and resplit the data! The new training points should be drawn from `X_train` and `y_train`.*
- Fit each model with each training set size and make predictions on the test set (9 in total).  
**Note:** Three tables are provided after the following code cell which can be used to store your results.

In [17]:
# TODO: Import the three supervised learning models from sklearn
# from sklearn import model_A
# from sklearn import model_B
# from skearln import model_C
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn import svm

# TODO: Initialize the three models
clf_A = DecisionTreeClassifier(random_state=0);
clf_B = GaussianNB()
clf_C = svm.SVC(random_state=0)

for clf in [clf_A,clf_B,clf_C]:

    # show the estimator type
    print "\n{}: \n".format(clf.__class__.__name__)

    # TODO: loop thru training sizes
    for n in [100,200,300]:

        # create training data from first 100, then 200, then 300 points
        train_predict(clf, X_train[:n], y_train[:n], X_test, y_test)





DecisionTreeClassifier: 

Training a DecisionTreeClassifier using a training set size of 100. . .
Trained model in 0.0027 seconds
Made predictions in 0.0007 seconds.
F1 score for training set: 1.0000.
Made predictions in 0.0003 seconds.
F1 score for test set: 0.7818.
Training a DecisionTreeClassifier using a training set size of 200. . .
Trained model in 0.0024 seconds
Made predictions in 0.0008 seconds.
F1 score for training set: 1.0000.
Made predictions in 0.0003 seconds.
F1 score for test set: 0.7843.
Training a DecisionTreeClassifier using a training set size of 300. . .
Trained model in 0.0032 seconds
Made predictions in 0.0008 seconds.
F1 score for training set: 1.0000.
Made predictions in 0.0003 seconds.
F1 score for test set: 0.7027.

GaussianNB: 

Training a GaussianNB using a training set size of 100. . .
Trained model in 0.0011 seconds
Made predictions in 0.0005 seconds.
F1 score for training set: 0.7969.
Made predictions in 0.0005 seconds.
F1 score for test set: 0.8148.
Tr

### Tabular Results
Edit the cell below to see how a table can be designed in [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables). You can record your results from above in the tables provided.

###### Decision Trees 
###### Random_state=0

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |             0.0027      |     0.0003             |  1.0000          |     0.7818      |
| 200               |        0.0025           |     0.0004             |  1.0000          |     0.7843      |
| 300               |        0.0033           |     0.0003             |  1.0000          |    0.7027      |

###### Naive Bayes

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |    0.0012           |       0.0005                 |   0.7969        |    0.8148           |
| 200               |     0.0012             |    0.0005                |    0.8030        |   0.7895         |
| 300               |     0.0013             |     0.0005                   |   0.7813     |     0.7965     |


###### SVM
###### Random_state=0

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |     0.0327              |   0.0013               | 0.8280           |    0.8154       |
| 200               |     0.0067              |  0.0022                |  0.8516          |    0.7937       |
| 300               |     0.0136              |  0.0029                |   0.8534            |    0.7937       |



## Choosing the Best Model
In this final section, you will choose from the three supervised learning models the *best* model to use on the student data. You will then perform a grid search optimization for the model over the entire training set (`X_train` and `y_train`) by tuning at least one parameter to improve upon the untuned model's F<sub>1</sub> score. 

### Question 3 - Choosing the Best Model
*Based on the experiments you performed earlier, in one to two paragraphs, explain to the board of supervisors what single model you chose as the best model. Which model is generally the most appropriate based on the available data, limited resources, cost, and performance?*

Based on above data both Decision tree & Naive Bayes seems good candidate to classify the student_data, based on training time & final F1_score on the test data. I chose decision tree over Naive bayes as the output model of decion tree is much intutive and can be used for understanding the data characteristics. One observation is decision tree with deafult hyperparameters seems to be overfitting the data, as F1_score of training data is quite high & there is divergence in the test & training score. I assume this will be corrected, if I further experiment with optimal parameters for Decision Trees. 

SVM are quite costly in terms of training time, so not good candidate for limited computation environment. 

### Question 4 - Model in Layman's Terms
*In one to two paragraphs, explain to the board of directors in layman's terms how the final model chosen is supposed to work. Be sure that you are describing the major qualities of the model, such as how the model is trained and how the model makes a prediction. Avoid using advanced mathematical or technical jargon, such as describing equations or discussing the algorithm implementation.*

Decision trees models are used for classifying the discrete data. The resulting model is in terms of if else statements that provides intutive details about characteristics of underlying data. The sequence of if-else decisions are represented in the decision tree. 

For the student_data set the model could be as simple as dividing

```python
if (number of hours student study per day > some threshold)
  classify as pass
else  
  classify as fail
```

The classifier works on given training data by chosing the best sequence in which attributes need to be checked to classify the data. The attribute selection is guided by the principle of maximizing information gain. Simply said we chose attributes which best seperates out the positive & negative samples. 

The classifier suffers from overfitting to the training data. In the worst case each leaf node would be containing single training data. This could be avoided by putting a constraint on the depth of tree or pruning the tree after the data classification is completed. Best method of avoiding overfitting is pruning the tree while testing it with test data,using the pruned tree which gives maximum test score.


The classification is simple tree traversal along sequence of if-else statements from root to leaf.The output of model in sklearn can also be dumped as pdf, giving tree structure, attribute sequence used in classification with all detils of informatin gain & mean square error at each node.


### Implementation: Model Tuning
Fine tune the chosen model. Use grid search (`GridSearchCV`) with at least one important parameter tuned with at least 3 different values. You will need to use the entire training set for this. In the code cell below, you will need to implement the following:
- Import [`sklearn.grid_search.gridSearchCV`](http://scikit-learn.org/stable/modules/generated/sklearn.grid_search.GridSearchCV.html) and [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Create a dictionary of parameters you wish to tune for the chosen model.
 - Example: `parameters = {'parameter' : [list of values]}`.
- Initialize the classifier you've chosen and store it in `clf`.
- Create the F<sub>1</sub> scoring function using `make_scorer` and store it in `f1_scorer`.
 - Set the `pos_label` parameter to the correct value!
- Perform grid search on the classifier `clf` using `f1_scorer` as the scoring method, and store it in `grid_obj`.
- Fit the grid search object to the training data (`X_train`, `y_train`), and store it in `grid_obj`.

In [20]:

# TODO: Import 'GridSearchCV' and 'make_scorer'
from sklearn.cross_validation import cross_val_score
from sklearn.grid_search import GridSearchCV
from sklearn.metrics import r2_score
from sklearn.metrics.scorer import make_scorer
from sklearn.cross_validation import ShuffleSplit

parameters = [{
    'max_depth': np.arange(3,30,1),
     'max_leaf_nodes' : np.arange(30,60,1)
      }]

# TODO: Initialize the classifier
clf = DecisionTreeClassifier()
cv_sets = ShuffleSplit(X_train.shape[0], n_iter = 10, test_size = 0.20, random_state = 0)

# TODO: Make an f1 scoring function using 'make_scorer' 
f1_scorer = make_scorer(f1_score, pos_label=1)


# TODO: Perform grid search on the classifier using the f1_scorer as the scoring method
grid_obj = GridSearchCV(estimator=clf, param_grid=parameters, cv=cv_sets ,scoring=f1_scorer)

# TODO: Fit the grid search object to the training data and find the optimal parameters
grid_obj = grid_obj.fit(X_train, y_train)

# Get the estimator
clf = grid_obj.best_estimator_
print clf
# Report the final F1 score for training and testing after parameter tuning
print "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
print "Tuned model has a testing F1 score of {:.4f}.".format(predict_labels(clf, X_test, y_test))


DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=3,
            max_features=None, max_leaf_nodes=47, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            presort=False, random_state=None, splitter='best')
Made predictions in 0.0006 seconds.
Tuned model has a training F1 score of 0.8500.
Made predictions in 0.0003 seconds.
Tuned model has a testing F1 score of 0.8065.


### Question 5 - Final F<sub>1</sub> Score
*What is the final model's F<sub>1</sub> score for training and testing? How does that score compare to the untuned model?*

1. Final training score is 0.8310
2. Final testing score is 0.8095

The test score with tuned parameters is quite close to trainign score. This gives confidence that tree is not being overfitted as earlier with default hyperparameters where there was hugh divergence between training & test scores. 



> **Note**: Once you have completed all of the code implementations and successfully answered each question above, you may finalize your work by exporting the iPython Notebook as an HTML document. You can do this by using the menu above and navigating to  
**File -> Download as -> HTML (.html)**. Include the finished document along with this notebook as your submission.